server/src/main/java/org/opensearch/gateway/TransportNodesBulkListGatewayStartedShards.java

```diff
@@ -0,0 +1,429 @@
+        List<FailedNodeException> failures
+    ) {
+        return new NodesGatewayStartedShards(clusterService.getClusterName(), responses, failures);
+    }
+
+    /**
+     * This function is similar to nodeoperation method of {@link TransportNodesListGatewayStartedShards} we loop over
+     * the shards here to fetch the shard result in bulk.
+     *
+     * @param request
+     * @return BulkOfNodeGatewayStartedShards
+     */
+    @Override
+    protected BulkOfNodeGatewayStartedShards nodeOperation(NodeRequest request) {
+        Map<ShardId, NodeGatewayStartedShards> shardsOnNode = new HashMap<>();
+        for (Map.Entry<ShardId, String> shardToCustomDataPathEntry : request.shardIdsWithCustomDataPath.entrySet()) {
+            try {
+                final ShardId shardId = shardToCustomDataPathEntry.getKey();
+                logger.trace("{} loading local shard state info", shardId);
+                ShardStateMetadata shardStateMetadata = ShardStateMetadata.FORMAT.loadLatestState(
+                    logger,
+                    namedXContentRegistry,
+                    nodeEnv.availableShardPaths(shardId)
+                );
+                if (shardStateMetadata != null) {
+                    if (indicesService.getShardOrNull(shardId) == null) {
+                        final String customDataPath = TransportNodesGatewayStartedShardHelper.getCustomDataPathForShard(
+                            logger,
+                            shardId,
+                            shardToCustomDataPathEntry.getValue(),
+                            settings,
+                            clusterService
+                        );
+                        // we don't have an open shard on the store, validate the files on disk are openable
+                        Exception shardCorruptionException = TransportNodesGatewayStartedShardHelper.getShardCorruption(
+                            logger,
+                            nodeEnv,
+                            shardId,
+                            shardStateMetadata,
+                            customDataPath
+                        );
+                        if (shardCorruptionException != null) {
+                            String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                            shardsOnNode.put(
+                                shardId,
+                                new NodeGatewayStartedShards(allocationId, shardStateMetadata.primary, null, shardCorruptionException)
+                            );
+                            continue;
+                        }
+                    }
+                    logger.debug("{} shard state info found: [{}]", shardId, shardStateMetadata);
+                    String allocationId = shardStateMetadata.allocationId != null ? shardStateMetadata.allocationId.getId() : null;
+                    final IndexShard shard = indicesService.getShardOrNull(shardId);
```

Comment:
Duplicate code ? 

Can't we extract full code inside for loop in helper ?

Reply:
Modified the helper method to move the entire code of fetching the shard information